File: Go/Kex_1sub.go
package main

import (
	"fmt"
	"math/rand"
	"strconv"
)

const AttackHit int = 110
const SkillHit int = 100
const MagicHit int = 70
const GolemHit int = 70
const GolemFlee int = 20

func checkHitOrMiss(hit int) bool {
	val := rand.Intn(100)
	return val <= hit - 1
}

func AttackGolemEx(lv int, player *Player) {
	golem := GolemEx{
		Hp:      lv * 50 + 100,
		Defense: lv * 10 + 40,
		Attack:  lv * 10 + 40,
		Hit:     GolemHit,
		Flee:    GolemFlee,
	}

	fmt.Printf("ゴーレムLv.%dが現れた！\n", lv + 1)
	for golem.Hp > 0 {
		fmt.Printf("ゴーレム 残りHP：%d\n", golem.Hp)
		fmt.Print("武器を選択してください（１．攻撃　２．特技　３．魔法）＞")
		var reader string
		_, _ = fmt.Scanln(&reader)
		choice, _ := strconv.Atoi(reader)
		var damage int
		var is_hit bool
		switch choice {
		case 1:
			damage = 60 + rand.Intn(41)
			is_hit = checkHitOrMiss(AttackHit - golem.Flee)
		case 2:
			damage = 30 + rand.Intn(101)
			is_hit = checkHitOrMiss(SkillHit - golem.Flee)
		case 3:
			damage = 20 + rand.Intn(181)
			is_hit = checkHitOrMiss(MagicHit - golem.Flee)
		default:
			damage = 0
		}

		if is_hit {
			damage -= golem.Defense
			if damage <= 0 {
				damage = 0
			}
			fmt.Printf("%dのダメージ！\n", damage)
			golem.Hp -= damage
		} else {
			fmt.Println("攻撃を外した！")
		}

		fmt.Println("ゴーレムの攻撃！")
		is_hit = checkHitOrMiss(golem.Hit)
		if is_hit {
			injury := golem.Attack - player.Defense
			fmt.Printf("%dのダメージ！\n", injury)
			player.Hp -= injury
			if player.Hp <= 0 {
				player.Hp = 0
				fmt.Println("あなたはゴーレムに負けました！ゲームオーバー！")
				return
			}
		} else {
			fmt.Println("攻撃を外した！")
		}

		fmt.Printf("プレーヤー残りHP：%d\n", player.Hp)
		if golem.Hp <= 0 {
			golem.Hp = 0
		}
		if golem.Hp == 0 {
			fmt.Printf("ゴーレムLv.%dを倒した！\n", lv + 1)
		}
	}
}

File: rust/src/kex_1sub.rs
use crate::golem::GolemEx;
use crate::player::Player;
use crate::read_user_input::read_user_input;
use rand::Rng;
use std::io::Write;

const ATTACK_HIT: i32 = 110;
const SKILL_HIT: i32 = 100;
const MAGIC_HIT: i32 = 70;

fn check_hit_or_miss(hit: i32) -> bool {
    let val = rand::thread_rng().gen_range::<i32>(0, 100);
    val <= hit - 1
}

pub fn attack_golem(lv: i32, player: &mut Player) {
    let mut golem = GolemEx::new(
        lv * 50 + 100,
        lv * 10 + 40,
        lv * 10 + 40
    );

    println!("ゴーレムLv.{}が現れた！", lv + 1);
    while golem.hp > 0 {
        println!("ゴーレム 残りHP：{}", golem.hp);
        print!("武器を選択してください（１．攻撃　２．特技　３．魔法）＞");
        std::io::stdout().flush().unwrap();
        let mut reader = String::new();
        let choice = match read_user_input(&mut reader) {
            Ok(x) => x,
            Err(e) => {
                eprintln!("エラーが発生しました：{}", e.to_string());
                return;
            }
        };
        let mut damage = 0;
        let mut is_hit = false;
        match choice {
            1 => {
                damage = 60 + rand::thread_rng().gen_range::<i32>(0, 41);
                is_hit = check_hit_or_miss(ATTACK_HIT - golem.flee);
            },
            2 => {
                damage = 30 + rand::thread_rng().gen_range::<i32>(0, 101);
                is_hit = check_hit_or_miss(SKILL_HIT - golem.flee);
            }
            3 => {
                damage = 20 + rand::thread_rng().gen_range::<i32>(0, 181);
                is_hit = check_hit_or_miss(MAGIC_HIT - golem.flee);
            },
            _ => ()
        }

        if is_hit {
            damage -= golem.defense;
            if damage <= 0 {
                damage = 0;
            }
            println!("{}のダメージ！", damage);
            golem.hp -= damage;
        } else {
            println!("攻撃を外した！");
        }

        println!("ゴーレムの攻撃！");
        is_hit = check_hit_or_miss(golem.hit);
        if is_hit {
            let injury = golem.attack - player.defense;
            println!("{}のダメージ！", injury);
            player.hp -= injury;
            if player.hp <= 0 {
                player.hp = 0;
                println!("あなたはゴーレムに負けました！ゲームオーバー！");
                return;
            }
        } else {
            println!("攻撃を外した！");
        }

        println!("プレーヤー残りHP：{}", player.hp);
        if golem.hp <= 0 { golem.hp = 0; }
        if golem.hp == 0 {
            println!("ゴーレムLv.{}を倒した！", lv + 1);
        }
    }
}